recipes/flatcc/all/test_package/CMakeLists.txt

```diff
@@ -1,32 +1,33 @@
-cmake_minimum_required(VERSION 2.8)
+cmake_minimum_required(VERSION 3.1)
 project(flatcc_example)
 
 include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
 conan_basic_setup()
 
-set(INC_DIR "${PROJECT_SOURCE_DIR}/include")
-set(GEN_DIR "${CMAKE_CURRENT_BINARY_DIR}/generated")
-set(FBS_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
 
-include_directories("${GEN_DIR}" "${INC_DIR}")
+set(FBS_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
 
-add_executable(monster monster.c)
 
 #On MacOS System Integrity Protection (SIP) will clear the DYLD_LIBRARY_PATH variable.
 #As a result calling flatcc from cmake will currently not work if the flatcc executable
 # is linked shared. As a workaround we generate the flatbuffer C files in the Conan recipe
 # when on MacOS and flatcc option 'shared' is True.
-if (NOT MACOS_SIP_WORKAROUND)
-    add_custom_target(gen_monster_fbs ALL)
-    add_custom_command (
-        TARGET gen_monster_fbs
-        COMMAND cmake -E make_directory "${GEN_DIR}"
-        COMMAND flatcc -a -o "${GEN_DIR}" "${FBS_DIR}/monster.fbs"
-        DEPENDS flatcc "${FBS_DIR}/monster.fbs"
+if (MACOS_SIP_WORKAROUND)
```

Comment:
Make sure that cmake in `test_package/conanfile.py` is only run when `not tools.cross_building(self.settings)` is true.

Reply:
Alternative is adding flatcc as build requirement in `test_package/conanfile.py`, but then when increasing the version in conanfile.py you have to remember to update the flatcc version in test_package/conanfile.py as well.

```
@@ -6,6 +6,7 @@ from conans import ConanFile, CMake, tools, RunEnvironment
 class FlatccTestConan(ConanFile):
     settings = "os", "compiler", "build_type", "arch"
     generators = "cmake"
+    build_requires = "flatcc/0.7.0pre"
```

Reply:
Wait, that won't work because the flatcc version in build_requires should be the same as the injected flatcc requirement.

Reply:
The test_package (=running `flatcc::cli`) only works when not cross building.
I haven't tested, but maybe add
```python
def build_requirements(self):
    if tools.cross_building(self.settings):
        self.build_requires("flatcc/{}"/format(some_version)
```

Reply:
But I don't think conan supports this.
I haven't seen recipes doing this. Maybe ask @jgsogo, the conan xbuild professional.

Reply:
I already tried to retrieve the flatcc version that is tested from self.requires (since Conan injects it for the test_package). But both in config_options() and configure() self.requires is empty.

I'll ask @jgsogo , else the following works when not using strange versions like 0.7.0pre:
`build_requires = ("flatcc/[>=0.7.0]")`

Reply:
That might work, but is not really reproducible.
There should be some way to know what package a test recipe is testing.
e.g. as `self.test_package`

Reply:
Not a big fan of this interface, but it is documented: you can use `self.deps_cpp_info["requirements"].version` in the `build()` to get the version of the requirement (after the graph is resolved). Is this what you are looking for? 

Conan injects the `requirement` automatically, but after that line, it is a regular requirement.

Reply:
Hi @jgsogo ,

What we are looking for is finding out the version of the package that is being tested at a point where we can inject a build_requirement for the same package/version.

In order to build the test_package for flatcc (and similar tools) when cross-compiling, flatcc is also needed in the build architecture to generate the flatbuffer header files. Then the test executable is linked to the flatcc host arch library.

I suppose in the build() function it is too late to inject a build_requirement. Is it possible to do that earlier ?

Reply:
I see.

This is not possible today. The recipe in the `test_package` works like any other recipe. The version of a requirement can be overridden by other branches of the graph (overrides, version-ranges, diamonds), so the final version of a requirement is only known when the graph is fully resolved and this is exactly the scenarios where the new `validate()` function plays an important role.

BUT, I agree this is a `test_package` and it totally makes sense to know the recipe being tested.  And probably Conan should forbid modifying options in the `test_package` too (https://github.com/conan-io/conan/issues/7547). These (together with a proposal to test build-requires using the two-profiles approach https://github.com/conan-io/conan/issues/7132) are things that we should explore before Conan v2.

Would you mind opening an issue with this feature request?

Meanwhile, I cannot see a documented way to achieve this behavior.